lib/browser-specific.js

```diff
@@ -0,0 +1,247 @@
+  while (browserTests.every(tests => tests.hasNext())) {
+    const firstTests = browserTests[0];
+
+    // If the tests are all the same object (which happens due to the caching in
+    // lib/results.js), we can just skip them; it is impossible for there to be
+    // browser-specific failures.
+    if (browserTests.every(t => t.currentValue() == firstTests.currentValue())) {
+      browserTests.forEach(t => t.next());
+      continue;
+    }
+
+    if (browserTests.every(t => t.currentKey() == firstTests.currentKey())) {
+      scoreTest(browserTests.map(t => t.currentValue()), scores);
+      browserTests.forEach(t => t.next());
+      continue;
+    }
+
+        // Our iterators are not pointing at the same test; find the earliest
+    // iterator and move it forward.
+    let smallestKey = firstTests.currentKey();
+    let smallestIdx = 0;
+    for (let i = 1; i < browserTests.length; i++) {
+      if (browserTests[i].currentKey() < smallestKey) {
+        smallestKey = browserTests[i].currentKey();
+        smallestIdx = i;
+      }
+    }
+    browserTests[smallestIdx].next();
+  }
+
+  // Now recurse into subtrees.
+  const browserSubtrees = browserTrees.map(tree => new IteratorHelper(tree.trees));
+  while (browserSubtrees.every(subtree => subtree.hasNext())) {
+    const firstTree = browserSubtrees[0];
+
+    // If the subtrees are all the same object (which happens due to the caching
+    // in lib/results.js), we can just skip them; it is impossible for there to
+    // be browser-specific failures in the subtree.
+    if (browserSubtrees.every(s => s.currentValue() == firstTree.currentValue())) {
+      browserSubtrees.forEach(s => s.next());
+      continue;
+    }
+
+        if (browserSubtrees.every(s => s.currentKey() == firstTree.currentKey())) {
+      walkTrees(browserSubtrees.map(s => s.currentValue()), scores);
+      browserSubtrees.forEach(s => s.next());
+      continue;
+    }
+
+    // Our iterators are not pointing at the same subtree; find the earliest
+    // iterator and move it forward.
+    let smallestKey = firstTree.currentKey();
+    let smallestIdx = 0;
+    for (let i = 1; i < browserSubtrees.length; i++) {
+      if (browserSubtrees[i].currentKey() < smallestKey) {
+        smallestKey = browserSubtrees[i].currentKey();
+        smallestIdx = i;
+      }
+    }
+    browserSubtrees[smallestIdx].next();
+  }
+}
+
+// Produces a 'score' of browser-specific failures for a given set of runs from
+// different products on the same WPT codebase. The word 'score' is used instead
+// of count as we normalize the counts of subtests.
+//
+// runs: an array of run objects, where each run has the form:
+//       {browser_name: "foo", tree: <an in-memory git tree>}
+//
+// expectedBrowsers: the set of browsers that should be (exactly) represented in
+//                   runs. If a browser is missing, an exception will be thrown.
+//
+// Returns a map from product name to score.
```

Comment:
Optional nit: The code will need to appear once regardless, but FWIW it would seem like a clearer interface to let `scoreBrowserSpecificFailures` just take an array of returns and return an array of scores. The calling code can worry about what browsers those runs are for and turning the scores into the right structure for output.